src/crate/client/converter.py

```diff
@@ -0,0 +1,147 @@
+# this file to you under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.  You may
+# obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+# However, if you have executed another commercial license agreement
+# with Crate these terms will supersede the license and you may use the
+# software solely pursuant to the terms of the relevant commercial agreement.
+"""
+Machinery for converting CrateDB database types to native Python data types.
+
+https://crate.io/docs/crate/reference/en/latest/interfaces/http.html#column-types
+"""
+import ipaddress
+from copy import deepcopy
+from datetime import datetime
+from enum import Enum
+from typing import Any, Callable, Dict, Optional, Union, List
+
+InputVal = Any
+
+
+def _to_ipaddress(value: Optional[str]) -> Optional[Union[ipaddress.IPv4Address, ipaddress.IPv6Address]]:
+    """
+    https://docs.python.org/3/library/ipaddress.html
+    """
+    if value is None:
+        return None
+    return ipaddress.ip_address(value)
+
+
+def _to_datetime(value: Optional[float]) -> Optional[datetime]:
+    """
+    https://docs.python.org/3/library/datetime.html
+    """
+    if value is None:
+        return None
+    return datetime.utcfromtimestamp(value / 1e3)
+
+
+def _to_default(value: Optional[InputVal]) -> Optional[Any]:
+    return value
+
+
+# Symbolic aliases for the numeric data type identifiers defined by the CrateDB HTTP interface.
+# https://crate.io/docs/crate/reference/en/latest/interfaces/http.html#column-types
+class CrateDatatypeIdentifier(Enum):
+    NULL = 0
+    NOT_SUPPORTED = 1
+    CHAR = 2
+    BOOLEAN = 3
+    TEXT = 4
+    IP = 5
+    DOUBLE = 6
+    REAL = 7
+    SMALLINT = 8
+    INTEGER = 9
+    BIGINT = 10
+    TIMESTAMP = 11
+    OBJECT = 12
+    GEOPOINT = 13
+    GEOSHAPE = 14
+    UNCHECKED_OBJECT = 15
+    REGPROC = 19
+    TIME = 20
+    OIDVECTOR = 21
+    NUMERIC = 22
+    REGCLASS = 23
+    DATE = 24
+    BIT = 25
+    JSON = 26
+    CHARACTER = 27
+    ARRAY = 100
+
+
+# Map data type identifier to converter function.
+_DEFAULT_CONVERTERS: Dict[int, Callable[[Optional[InputVal]], Optional[Any]]] = {
+    CrateDatatypeIdentifier.IP.value: _to_ipaddress,
+    CrateDatatypeIdentifier.TIMESTAMP.value: _to_datetime,
+}
+
+
+class Converter:
+    def __init__(
+        self,
+        mappings: Dict[int, Callable[[Optional[InputVal]], Optional[Any]]] = None,
+        default: Callable[[Optional[InputVal]], Optional[Any]] = _to_default,
+    ) -> None:
+        self._mappings = mappings or {}
+        self._default = default
```

Comment:
Could promote usage of the enum further and use it in `mappings` as key instead of the `int` value.

That would implicitly document what `int` values are acceptable.

Reply:
Thank you for bringing this up.

### Problem
Without looking further into the details and the state of the onion, I was wary of using an Enum-type as key, because I found different resources like those referenced below, which indicate possible performance losses. Especially because the converter sits in a busy place of the driver, I decided to not use it this way and use an `int` as key instead.

### Thoughts
If we still want to use an Enum-type as a key on the mapping data structure itself, I can think of two options, but I will be happy to hear if you see others.
- Use in-memory `@memoize`-style caching on the lookup path, if that would work around the Enum-lookup-is-slow problem.
- Look into using the `fastenum` library.

### References

TLDR; Apparently, Python 3.4's first Enum implementation was pretty slow. While it got better with subsequent releases, the `fastenum` library still claims to be up to 3.5x faster in benchmarks on recent versions of Python.

- https://quantlane.com/blog/fastenum/, https://gitlab.com/quantlane/libs/fastenum
- https://pypi.org/project/fast-enum/
- https://bugs.python.org/issue39102

Reply:
Is this relevant for how we use it?

What we should do independent of this is to  move the dict lookup out of the row loop. The columns/types are the same across all rows. We can retrieve the converter for all columns once and then process all rows with them.

Then you've 1 enum attribute access and 1 dict lookup *per* column. That shouldn't be a performance issue.